### ISSUE TITLE
Improved Decoupling of API & UI

### ISSUE THREAD

**Issue by kahboom** [kahboom/syndesis#1]
## tl;dr
Following from our discussion in the retrospective, there is an inherent problem in the application development lifecycle that I think is directly impeding our ability to keep up with feature requests in a timely fashion. I know this is a LONG description, but please take a moment to read, as I think _everyone_ is affected by this. It's an opportunity to improve the development process. If we don't solve this soon, technical debt will continue to increase and the overall product roadmap will be jeopardized.

The two items to address are 1) decoupling the API & UI; 2) making a more parallel work flow rather than linear and blocking. Solutions proposed are at the bottom. We should collaborate and discuss any potential tradeoffs or alternatives.

## Introduction
At the beginning of the project we made a huge effort to decouple the UI from the backend by creating a Node.js API (now in Java). As features continue to be added to the iPaaS roadmap, we are finding it more and more challenging to develop these features without getting blocked by one or more areas.

##  Theory
When you are spending as much time trying to get the local build of the REST API working properly (to be able to work with real data) as you do actually implementing a feature, it's a problem. It almost feels like we need yet another layer of abstraction to be able to work completely decoupled from the backend. From the UI perspective, which sits kind of right in the middle, it boils down to two factors:

1) separation of concerns and true decoupling
2) timing per sprint to develop these features (linear vs parallel work flow)

## Current Work Process
Here is a rough outline of our work process:

1. Discuss requirements for the sprint.
2. Assign feature owners. Feature owners are usually backend folk (not sure if this is intentional, but seems like a very important consideration we should be consistent with and have a philosophy it supports).
3. Feature owners write a design proposal, which includes the data structure the UI should adhere to.
4. UXD collaborates with feature owners to create designs.
5. UI receives designs, looks at design proposal for the REST API, and begins work.
6. Confusion about whether to mock out data or use existing REST API work ensues. Neither option is smooth, as using existing resources is risky for reasons stated below, and mocked data is great, but when it comes time to wire up the API, there are often huge discrepencies.
7. Scrambling to get things working properly together, whether using maintainable approaches or not, in time for the demo and sprint end.

## Testing the Theory
The problems described earlier manifest in a few different ways that I've noticed impede development:

### Separation of Concerns
- The REST API sometimes has continuously changing requirements anywhere in steps 3-5 that result in adjustments to the data it feeds to the UI (and its structure), possible changes in endpoints, sometimes the structure is not known at all, etc. This is often out of necessity due to potential issues that are later discovered in the design proposal (naturally, this is much easier to discover on implementation rather than in theory on a PR, which in and of itself is a problem). Other times they are dealing with CI issues.
- The UI tries to satisfy design requirements but they sometimes do not match the data structure provided by the API. Improvements have been made here by increasing communication between UXD and the backend, but sometimes, as stated earlier, structural issues and discrepencies arise mid-implementation, which is normal. Backend issues are discovered, such as a bug or server issues with using the REST API. Also, a lot of the data returned from the API is not relevant and the structure is inconsistent across features.

### Timing
- The REST API & UXD teams are essentially trying to work in parallel with the UI team, which puts a lot of pressure on the former teams to get things completed in time for implementation in the UI, and early enough for QE and Docs to also make adjustments (a deadline the UI often can't meet). Parallel should be our goal, but execution is often linear and blocking.
- Since much of the sprint time in the UI is spent creating components and templates for the design implementation, this is often discovered at the very end of the sprint when wiring up the API.

The UI should be able to operate strictly with a known data structure and rough designs. Whether or not that has been implemented in the REST API is irrelevant. The data structure should be known before a single line of code is written on either side.

But who dictates what the structure should be? At the moment it is usually the feature owner, which is usually a backend engineer. There is nothing wrong with that, but if we choose to go that route, some adjustments need to be made.

## Next Steps
Some proposed solutions (feel free to add):

- Decide where the data structure should come from. Considering that iPaaS is heavily user-focused and technically SaaS, to me, it makes sense that it either comes completely from the front end or is mediated between the API and UI (more on that below).
- Decide how that data structure is decided on:
  - Consumer-driven contracts, as @zregvart  suggested. ([Concept](https://martinfowler.com/articles/consumerDrivenContracts.html) & [example](https://cloud.spring.io/spring-cloud-contract/) for Spring). A pattern to adhere to an evolving service, and can come in many forms, as simple as a spreadsheet.
  - UI-driven data models, to which the API then adapts. This doesn't mean UI engineers need to be feature owners, but just that the UI portion would dictate the structure. This would allow us to mock out data knowing that when we switch to the REST API it will be a seamless transition (in theory).
  - Using an intermediary solution, like [GraphSQL](http://graphql.org/) and [Apollo](https://www.apollographql.com/client/) (here is a full stack [tutorial](https://www.howtographql.com/)). Not a huge fan, but it could work.
  - Putting a freeze on design proposals, regardless of how bad the structure is, and reconvening the next sprint?

@syndesisio/all - Let's discuss benefits and tradeoffs, other possible solutions, and any considerations below!

**Comment by deeleman** [comment]
For me the way forward is crystal clear:

>  Considering that iPaaS is heavily user-focused and technically SaaS, to me, it makes sense that it either comes completely from the front end or is mediated between the API and UI

and...

> UI-driven data models, to which the API then adapts. This doesn't mean UI engineers need to be feature owners, but just that the UI portion would dictate the structure. This would allow us to mock out data knowing that when we switch to the REST API it will be a seamless transition

Are the way to go IMO. 

I point this out based on how I've been working over the past years with large Angular, Backbone and Knockout enterprise applications consuming microservices and this has worked extremely well for me. On the contrary, every time the UI/UX setup had to conform to a data design forced by the backend layer, the end user experience turned into a fiasco most of the times.

Bear in mind that modern UI applications are not mere presentation layers anymore. Modern SPAs deliver nowadays complex navigational and interactive flows that also rely on functional, persisted, reactive data models and, as such, UI should take the initiative in design of the API interfaces. Conversations between UXD and UI will lead to leaner data experiences and a smoother integration between data logic, interaction logic and business logic (clientside).

**Comment by kahboom** [comment]
Thanks for the great feedback @deeleman , feel free to copy and paste in the new issue. 😂
